lib/bibliothecary/parsers/maven.rb

```diff
@@ -172,30 +174,53 @@ def self.extract_pom_info(xml, location, parent_properties = {})
       def self.extract_pom_dep_info(xml, dependency, name, parent_properties = {})
         field = dependency.locate(name).first
         return nil if field.nil?
+
         value = field.nodes.first
-        match = value.match(/^\$\{(.+)\}/)
+        match = value.match(MAVEN_PROPERTY_REGEX)
         if match
-          # the xml root is <project> so lookup the non property name in the xml
-          # this converts ${project/group.id} -> ${group/id}
-          non_prop_name = match[1].gsub('.', '/').gsub('project/', '')
-          return value if !xml.respond_to?('properties') && parent_properties.empty? && !xml.locate(non_prop_name)
-          prop_field = xml.properties.locate(match[1]).first
-          parent_prop = parent_properties[match[1]]
-          if prop_field
-            return prop_field.nodes.first
-          elsif parent_prop
-            return parent_prop
-          elsif xml.locate(non_prop_name).first
-            # see if the value to look up is a field under the project
-            # examples are ${project.groupId} or ${project.version}
-            return xml.locate(non_prop_name).first.nodes.first
-          else
-            return value
-          end
+          return extract_property(xml, match[2], value, parent_properties)
         else
           return value
         end
       end
+
+      def self.replace_value_with_prop(original_value, property_value, property_name)
+        original_value.gsub("${#{property_name}}", property_value)
+      end
+
+      def self.extract_property(xml, property_name, value, parent_properties = {}, depth = 0)
```

Comment:
I think if you break this into two methods it'll flow more easily, something like:
```ruby
      def self.extract_property(xml, property_name, value, parent_properties = {}, depth = 0)
        prop_value = property_value(xml, property_name, parent_properties)
        return value unless prop_value
        resolved_value = replace_value_with_prop(value, prop_value, property_name)

        # check to see if we just resolved to another property name
        match = resolved_value.match(MAVEN_PROPERTY_REGEX)
        if match && depth < MAX_DEPTH
          extract_property(xml, match[2], resolved_value, parent_properties, depth + 1)
        else
          resolved_value
        end
      end

      def self.property_value(xml, property_name, parent_properties)
        # the xml root is <project> so lookup the non property name in the xml
        # this converts ${project/group.id} -> ${group/id}
        prop_field = xml.respond_to?(:properties) && xml.properties.locate(property_name).first
        non_prop_name = property_name.gsub(".", "/").gsub("project/", "")

        if prop_field
          prop_field.nodes.first
        elsif parent_properties[property_name]
          parent_properties[property_name]
        elsif (non_prop = xml.locate(non_prop_name).first)
          # see if the value to look up is a field under the project
          # examples are ${project.groupId} or ${project.version}
          non_prop.nodes.first
        end

        # property not found
      end
```

Reply:
n.b. I did not test this